internal/jobqueue/dbjobqueue/dbjobqueue.go

```diff
@@ -0,0 +1,390 @@
+		if err != nil {
+			return uuid.Nil, uuid.Nil, nil, "", nil, fmt.Errorf("error waiting for notification on jobs channel: %v", err)
+		}
+	}
+
+	// insert heartbeat
+	_, err = conn.Exec(ctx, sqlInsertHeartbeat, token, id)
+	if err != nil {
+		return uuid.Nil, uuid.Nil, nil, "", nil, fmt.Errorf("error inserting the job's heartbeat: %v", err)
+	}
+
+	dependencies, err := q.jobDependencies(ctx, conn, id)
+	if err != nil {
+		return uuid.Nil, uuid.Nil, nil, "", nil, fmt.Errorf("error querying the job's dependencies: %v", err)
+	}
+
+	return id, token, dependencies, jobType, args, nil
+}
+
+func (q *dbJobQueue) FinishJob(id uuid.UUID, result interface{}) error {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return fmt.Errorf("error connecting to database: %v", err)
+	}
+	defer conn.Release()
+
+	// todo return errnotexist as well if it can't find it
+	// TODO check if job is not running first and return errnotrunning or errcanceled
+
+	var finished *time.Time
+	canceled := false
+	err = conn.QueryRow(context.Background(), sqlQueryJob, id).Scan(nil, nil, nil, &finished, &canceled)
+	if err == pgx.ErrNoRows {
+		return jobqueue.ErrNotExist
+	}
+	if canceled {
+		return jobqueue.ErrCanceled
+	}
+	if finished != nil {
+		return jobqueue.ErrNotRunning
+	}
+
+	tx, err := conn.Begin(context.Background())
+	if err != nil {
+		return fmt.Errorf("error starting database transaction: %v", err)
+	}
+	defer func() { _ = tx.Rollback(context.Background()) }()
+
+	// Remove from heartbeats
+	tag, err := conn.Exec(context.Background(), sqlDeleteHeartbeat, id)
+	if err != nil {
+		return fmt.Errorf("error finishing job %s: %v", id, err)
+	}
+
+	if tag.RowsAffected() != 1 {
+		return jobqueue.ErrNotExist
+	}
+
+	tag, err = conn.Exec(context.Background(), sqlFinishJob, result, uuid.Nil, id)
+	if err != nil {
+		return fmt.Errorf("error finishing job %s: %v", id, err)
+	}
+
+	if tag.RowsAffected() != 1 {
+		return jobqueue.ErrNotExist
+	}
+
+	err = tx.Commit(context.Background())
+	if err != nil {
+		return fmt.Errorf("unable to commit database transaction: %v", err)
+	}
+
+	return nil
+}
+
+func (q *dbJobQueue) CancelJob(id uuid.UUID) error {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return fmt.Errorf("error connecting to database: %v", err)
+	}
+	defer conn.Release()
+
+	tag, err := conn.Exec(context.Background(), sqlCancelJob, id)
+	if err != nil {
+		return fmt.Errorf("error canceling job %s: %v", id, err)
+	}
+
+	if tag.RowsAffected() != 1 {
+		return jobqueue.ErrNotRunning
+	}
+
+	return nil
+}
+
+func (q *dbJobQueue) JobStatus(id uuid.UUID) (result json.RawMessage, queued, started, finished time.Time, canceled bool, deps []uuid.UUID, err error) {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return
+	}
+	defer conn.Release()
+
+	var sp, fp *time.Time
+	var rp pgtype.JSON
+	err = conn.QueryRow(context.Background(), sqlQueryJobStatus, id).Scan(&rp, &queued, &sp, &fp, &canceled)
```

Comment:
Doesn't it pass double pointers inside? (`**time.Time`) Is that what's desired? Feels weird.

Reply:
huh, I'll try without

Reply:
ugh I somehow messed this up and replied to the wrong comment: https://github.com/osbuild/osbuild-composer/pull/1539#discussion_r672175785

Reply:
Hm so if we change it, we'd need to have time.Time default-values stored, so I think the current form is fine from that perspective. 

Then again I do the opposite for uuids, where I do use `uuid.Nil`.

What do you prefer?

Reply:
if time can be `nil`, this completely makes sense then. Maybe put there a comment? Something like "we're passing a double pointer here so the method can return nil".

Reply:
Will do!